### DISCUSSION TITLE
How would one implement a conveyor belt with Jolt?

### DISCUSSION THREAD

**Discussion by SirLynix** [jrouwe/JoltPhysics#531] [General]
Hi,

I've been playing a bit with Jolt Physics and wondering how to implement a conveyor belt which would work with every rigid body, Chipmunk Physics (2D physics engine) has "surface velocity" property for that (which was useful also for character controller) but I think Jolt only has support for custom velocity for CharacterVirtual?

I can only think of two other ways of doing this, either a volume trigger to collect bodies on top of the conveyor belt (not very precise) or a specific material type so one can detect when a body start touching the conveyor surface (and stop touching it also).

However, how would one implement a proper conveyor belt? One that can handle objects partially sitting on it? (for example, a box with only one edge on the belt and one edge on something static wouldn't just acquire velocity, it would also rotate).

What do you think would be the best way to handle this with Jolt?

Thank you

**Comment by jrouwe** [comment]
The proper solution would be to create a conveyor belt out of multiple kinematic bodies that move as the segements of the conveyor belt would, but if you only want to use one body you could create a ```EMotionType::Kinematic``` body with ```EMotionQuality::Discrete``` for the conveyor belt and give the object the velocity that you want to give your conveyor belt (assuming there are no turns in the belt). Then at the end of each time step you would position the object back to its starting position using ```BodyInterface::SetPosition```. The position update of the object happens very late in the simulation step, so the fact that the conveyor belt body moved should not affect the other bodies (unless you're using multiple collision or integration steps in ```PhysicsSystem::Update```).

Note that a flag could be added to e.g. ```Body::EFlags``` that would turn off integration for the body in ```PhysicsSystem::JobIntegrateVelocity``` which would mean you don't need to put the object back as it won't move. But that feels very hacky. There are other alternatives, but they cost memory and performance in the contact solver which is a hot spot, so I'm hesitant to do this.

**Reply by SirLynix** [reply]
Sorry, I forgot to reply.

Thanks for the tip. However "moving" the physical body is not really something I could do (or I'd need to have a lot of them).

A way to set a custom contact joint velocity for custom materials would be great to handle this kind of stuff (and others, basically allowing a `CharacterContactListener::OnAdjustBodyVelocity` without requiring a character).

**Reply by jrouwe** [reply]
Unfortunately it's not as simple as that. The velocity is fetched multiple times during the solver update (because it changes) so it would incur a lot of overhead to the simulation.

**Comment by jrouwe** [comment]
This is implemented with #551

**Reply by SirLynix** [reply]
Awesome! Thank you!
